File: ruby/lib/redis/commands/publish.rb
require_relative './base'

module Redis
  module Commands
    class Publish < Base
      def self.command
        'publish'
      end

      def run
        channel_name = args[0]
        message = args[1]

        sockets = []
        @mutex.synchronize do
          sockets = @subscribed_channels[channel_name]&.dup || []
        end

        count = 0
        sockets.each do |socket|
          begin
            payload = Redis::Resp::Serializer.serialize(['message', channel_name, message])
            socket.write(payload)
            socket.flush
            count += 1
          rescue Errno::EPIPE, Errno::ECONNRESET, IOError => e
            puts "Failed to publish message: #{e.message}"
            @mutex.synchronize { @subscribed_channels[channel_name]&.delete(socket) }
          end
        end

        Redis::Resp::Serializer.serialize(count)
      end
    end
  end
end

File: python/src/commands/publish.py
from .base import BaseCommand
from src.resp.serializer import RespSerializer
import asyncio

class Publish(BaseCommand):
  @classmethod
  def command(cls) -> bytes:
    return b'publish'
  
  def run(self):
    channel_name: bytes = self.args[0]
    message: bytes = self.args[1]
    count: int = 0

    stream_writers: list[asyncio.StreamWriter] = self.subscribed_channels[channel_name] or []
    if channel_name and message:
      if stream_writers:
        payload = RespSerializer.serialize(data=[b'message', channel_name, message])
        for stream_writer in stream_writers:
          stream_writer.write(payload)
          asyncio.create_task(stream_writer.drain())
          count += 1
    
    return RespSerializer.serialize(data=count)
      